text/3414-deprecate-spooky-dropck-at-a-distance.md

```diff
@@ -0,0 +1,362 @@
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+  let mut v = vec![];
+  {
+    v.push(PrintOnDrop(&*String::from("temporary")));
+  }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC removes the dropck/outlives constraints from `PhantomData` and `[T; 0]`
+and moves them into the relevant `Drop` impls instead.
+
+Instead of relying on `PhantomData`, there are now 3 forms of `may_dangle`:
+
+For lifetime parameters: `#[may_dangle] 'a`, places no constraints on `'a`. This
+is unchanged from the current form.
+
+For type parameters that are dropped, they need to be annotated with
+`#[may_dangle(drop)]`. These type parameters will be checked as if in a struct
+like so:
+
+```rust
+struct Foo<T>(T);
+
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+  let mut foo;
+  {
+    foo = Foo(PrintOnDrop(&*String::from("temporary")));
+  }
+  // ERROR: Foo dropped here, runs destructor for PrintOnDrop
+}
+```
+
+For type parameters that are not dropped, `#[may_dangle(borrow)]` can be used.
+These are checked as if in a struct like so:
+
+```rust
+struct Foo<T>(*const T);
+
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+  let mut foo;
+  {
+    foo = Foo(&PrintOnDrop(&*String::from("temporary")));
+  }
+  // no error here
+}
+```
+
+Effectively, a type is checked to be *safe to drop* by the following procedure:
+
+- If the type has no type or lifetime parameters, it is *safe to drop*.
+- If the type is any of the below, it is *safe to drop*:
+    References, raw pointers, function pointers, function items, `PhantomData`,
+    `ManuallyDrop` and empty arrays. In other words, all (at the time of
+    writing) non-`needs_drop` types.
+- If the type has a `Drop` impl, or is a trait object:
+    - For every lifetime parameter:
+        - If the lifetime parameter is marked `#[may_dangle]`, continue.
+        - If the lifetime is dead, the type is not *safe to drop*.
+    - For every type parameter:
+        - If the type parameter is marked `#[may_dangle(borrow)]`, continue.
+        - If the type parameter is marked `#[may_dangle(drop)]`:
+            - If the type parameter is not *safe to drop*, then the type is not
+                *safe to drop*.
+            - Continue.
+        - If the type parameter contains lifetimes which are dead, the type is
+            not *safe to drop*.
+- If the type does not have a `Drop` impl:
+    - For every field:
+        - If the field type is not *safe to drop*, then the type is not *safe to
+            drop*.
+
+(N.B. you cannot add `#[may_dangle]` to a trait object's parameters.)
+
+This is different from the status quo in that 1. we skip checking fields
```

Comment:
What is very much missing from this RFC is some guidelines or suggestions about when to use the various forms. I see the changes to the algorithm, but I don't understand yet what structs should use `may_dangle(borrow)` vs `may_dangle(drop)`?